src/hotspot/share/prims/universalUpcallHandler.cpp

```diff
@@ -71,6 +72,93 @@ void ProgrammableUpcallHandler::detach_thread(Thread* thread) {
   vm->functions->DetachCurrentThread(vm);
 }
 
+// modelled after JavaCallWrapper::JavaCallWrapper
+Thread* ProgrammableUpcallHandler::on_entry(OptimizedEntryBlob::FrameData* context) {
+  JavaThread* thread = maybe_attach_and_get_thread(&context->should_detach)->as_Java_thread();
```

Comment:
Nit: if `maybe_attach_and_get_thread` has to return a JavaThread it should be typed to return a JavaThread.